# Protein Block Design

## About the project
Master's degree project aiming to design protein sequences folds using the "Protein Block" approach.


## Getting started

### Installation
1. Clone the repo:
```
git clone https://github.com/Damien-Garcia-Bioinformatics/protein-design.git
```
2. Install requirements:
```
pip install -r requirements.txt
```

#### Full requirements list
```
contourpy==1.0.7
cycler==0.11.0
fonttools==4.38.0
joblib==1.2.0
kiwisolver==1.4.4
matplotlib==3.6.3
numpy==1.24.2
packaging==23.0
pandas==1.5.3
Pillow==9.4.0
pyparsing==3.0.9
python-dateutil==2.8.2
pytz==2022.7.1
six==1.16.0
tqdm==4.64.1
tqdm-joblib==0.0.2
```

### Usage

### Roadmap

### License
Distributed under the GNU GENERAL PUBLIC LICENSE (Version 3, 29 June 2007).
See ```LICENSE``` for more information.

### Contact
- Author: Damien Garcia
- eMail: damien.garcia@etu.univ-nantes.fr
- Project link: https://github.com/Damien-Garcia-Bioinformatics/protein-design


## Acknowledgments
- **FORSA:**
    - Use of a structural alphabet to find compatible folds for amino acid sequences.
    *Swapnil Mahajan, Alexandre G. de Brevern, Yves-Henri Sanejouand, Narayanaswamy Srinivasan, Bernard Offmann*

- **Protein Blocks:**
    - A substitution matrix for structural alphabet based on structural alignment of homologous proteins and its applications
    *Manoj Tyagi, Venkataraman S. Gowri, Narayanaswamy Srinivasan, Alexandre G. de Brevern, Bernard Offmann*.
    - Extension of a local backbone description using a structural alphabet: A new approach to the sequence-structure relationship
    *Alexandre G. de Brevern, Hélène Valadié, Serge Hazout, Catherine Etchebest*